### PR TITLE
add ability to start container tests in debug mode and attach a debugger

### PR DESCRIPTION
**Pull request by dhiaayachi**
This PR add all the needed components to run consul test-containers tests in debug mode which enable the ability of attaching a debugger to consul while running those tests.

- Add an image that have dlv + consul
- Add a `debug` flag to the tests which will run consul with dlv and expose an extra port for debugging

### CODE REVIEW

File: test/integration/consul-container/libs/cluster/container.go [hunk 1]
```diff
@@ -156,6 +157,20 @@ func NewConsulContainer(ctx context.Context, config Config, cluster *Cluster, po
 
 		info AgentInfo
 	)
+	if utils.Debug {
+		for i := 0; i < 10; i++ {
```
Comment (huikang) [inline test/integration/consul-container/libs/cluster/container.go:161]: I guess we can replace this with `retry-go`, which is imported to the integ test:

https://github.com/hashicorp/consul/blob/44b39240a86bc94ddc67bc105286ab450bd869a9/test/integration/consul-container/libs/cluster/cluster.go#L141-L157

Comment (dhiaayachi) [reply test/integration/consul-container/libs/cluster/container.go]: Thank you for the review @huikang ! I changed that and also added few changes to print some data about the node when printing the debug port to help identify which container we would like to debug in a multi Consul test.

File: GNUmakefile [hunk 2]
```diff
@@ -163,6 +163,19 @@ dev-build:
 	rm -f ./bin/consul
 	cp ${MAIN_GOPATH}/bin/consul ./bin/consul
 
+
+dev-docker-dbg: dev-docker linux dev-build
+	@echo "Pulling consul container image - $(CONSUL_IMAGE_VERSION)"
+	@docker pull consul:$(CONSUL_IMAGE_VERSION) >/dev/null
+	@echo "Building Consul Development container - $(CONSUL_DEV_IMAGE)"
+	@#  'consul:local' tag is needed to run the integration tests
```
Comment (eculver) [inline GNUmakefile:171]:
```suggestion
	@#  'consul-dbg:local' tag is needed to run the integration tests
```

**Review by eculver — APPROVED** [review]
Looks pretty spectacular to me! Thanks for this :) 

I just have a couple non-blocking questions/comments.

File: test/integration/consul-container/libs/cluster/builder.go [hunk 3]
```diff
@@ -306,11 +306,15 @@ func (b *Builder) ToAgentConfig(t *testing.T) *Config {
 	confCopy, err := b.conf.Clone()
 	require.NoError(t, err)
 
+	cmd := []string{"agent"}
+	if utils.Debug {
+		cmd = []string{"/root/go/bin/dlv", "exec", "/bin/consul", "--listen=:4000", "--headless=true", "", "--accept-multiclient", "--continue", "--api-version=2", "--", "agent", "--config-file=/consul/config/config.json"}
```
[the remark this answers is not in the document]
Comment (dhiaayachi) [reply test/integration/consul-container/libs/cluster/builder.go:311]: yes at least for the base image. I think without dlv it's passed through env vars which dlv don't pass along to the sub process.

**Pull request merged**